### PR TITLE
Stability experiments : mutliple pseudobulks in the loss

### PR DESCRIPTION
**Pull request by SimonGrouard**
This PR was created to understand why the model metrics were not reproducible through different runs, which was a flaw to choose our hyperparameter values when they were not ordinal.
To do so, several changes were implemented:
1. Fix the train/val split seed to 42.
2. Sample several pseudobulks from a given batch in the training of MixUpVI, to show more than one pseudobulk per epoch (indeed, the classical ELBO loss is computed over every cell per batch, not just one, so maybe a similar idea for the pseudobulk loss would help it and make the model more stable). See below for more info.
3. Fix the torch seed to a value defined in `constants.py`. This allows to fix weight initialisation, the optimizer, and the sampling from the latent space. **Consequently, one can now tune the hyperparameters over several torch seeds, and plot the tuning graphs with a confidence interval due to the different seeds tried.**

To try and make the model more stable over different runs, and also as a general improvement idea, this PR allows to sample several pseudobulks in a given sample.. Therefore several attributes are now created and tunable : 
1. n_pseudobulks : how many pseudobulks are created in the batch
4. n_cells_per_pseudobulk : how many bootstrapped cells should every pseudobulk be composed of. Can be a maximum of n_batch cells.
5. mixup_penalty_aggregation : now there are as many penalty values as there are pseudobulks. How can to only get a scaler ? Either take the maximum value across pseudobulks, or the average, or the sum.
6. average_variables_mixup_penalty: whether to average (or sum if False) the penalty values across variables. Variables are just a few if the loss is computed in the latent space, therefore simply summing can be good. But inside the reconstructed space, variables are as many as n_input genes, so averaging can be better there.

_Finally, it is now possible to compute the L2 loss inside the reconstructed space, because instead of sampling a discrete distribution (making the reparametrization trick impossible), we just use the mean of the reconstructed distribution for the L2 loss._

### CODE REVIEW

File: scvi/dataloaders/_data_splitting.py [hunk 1]
```diff
@@ -223,7 +223,7 @@ def setup(
         n_val = self.n_val
         indices = np.arange(self.adata_manager.adata.n_obs)
 
-        random_state = np.random.RandomState(seed=settings.seed) # settings.seed is None so new random state for each run
+        random_state = np.random.RandomState(seed=42) # settings.seed is None so new random state for each run
```
Comment (SimonGrouard) [inline scvi/dataloaders/_data_splitting.py:226]: @khalilouardini  What do you think of always fixing the train/val split seed ?

Comment (khalilouardini) [reply scvi/dataloaders/_data_splitting.py:226]: For the purpose of the stability experiments it makes sense yes!

So that we can isolate the randomness that comes from the model.

File: scvi/module/_mixupvae.py [hunk 2]
```diff
@@ -157,11 +168,17 @@ def __init__(
...
         # MixUpVAE specific arguments
+        n_pseudobulks: Tunable[int] = 1,
+        n_cells_per_pseudobulk: Tunable[Optional[int]] = None,
         signature_type: Tunable[str] = "pre_encoded",
         loss_computation: Tunable[str] = "latent_space",
         pseudo_bulk: Tunable[str] = "pre_encoded",
         mixup_penalty: Tunable[str] = "l2",
+        mixup_penalty_aggregation: Tunable[str] = "mean",
+        average_variables_mixup_penalty: Tunable[bool] = False,
     ):
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py:182]: From these new hyperparameters, 

which ones will be included in the stability experiments ?

File: benchmark_utils/tuning_utils.py [hunk 3]
```diff
@@ -26,14 +24,31 @@ def format_and_save_tuning_results(tuning_results, variable: str, training_datas
                     else:
                         results[key].append(np.nan)
         results = pd.DataFrame(results)
-        results["hyperparameter"] = hyperparameter
+        
+        hyperparameters = path.split("/")[6]
```
Comment (khalilouardini) [inline benchmark_utils/tuning_utils.py]: I get this bug here:

image

Comment (khalilouardini) [reply benchmark_utils/tuning_utils.py]: If you want to subset the string with hyperparameters (not the most reliable way tbh) and expect to always get it from the last string after the split, I suggest using `.split("/")[-1]`

File: scvi/module/_mixupvae.py (hunk 2 above)
Comment (khalilouardini) [reply scvi/module/_mixupvae.py:182]: I feel like the list of hyperparameters is starting to become hard to manage. After the stability experiments I suggest we clean up the class and remove hyperparameters that remain fixed / haw negligible effect from the constants.

**Review by khalilouardini — COMMENTED** [review]
Thanks Simon! a few comments

File: scvi/module/_mixupvae.py (hunk 2 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py:182]: Not sure yet, I want to first understand the overall effect of having several pseudobulks in a same batch.
But for instance the `average_variables_mixup_penalty ` is set to False for the general use case, but I think would be better as True for the case where the MixUpLoss is computed in the reconstructed space, since otherwise it gets huge...
But agreed on the second comment

File: scvi/module/_mixupvae.py [hunk 4]
```diff
@@ -287,42 +313,57 @@ def inference(
...
+        for ground_truth in y_:
+            unique_indices, counts = ground_truth.unique(return_counts=True)
+            if len(counts) < self.n_labels:
+                # then not all labels are present in pseudobulk, but we need to specify these proportions of 0
+                unique_indices = unique_indices.int().tolist()
+                counts =  [counts[unique_indices.index(j)].item() if j in unique_indices else 0 for j in range(self.n_labels)]
+                counts = torch.tensor(counts).to(device="cuda")
+            proportions = counts.float() / n_cells_per_pseudobulk
+            all_proportions.append(proportions)
+
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py]: Maybe we should move this piece of code to a `mixupvi_utils.py` file in a `bootstrap_pseufobulk` function , the code is becoming harder to read !

File: scvi/module/_mixupvae.py [hunk 5]
```diff
@@ -287,42 +313,57 @@ def inference(
...
-        x_pseudobulk_ = x.mean(axis=0).unsqueeze(0)
 
-        # create signature
-        unique_indices, counts = y.unique(return_counts=True)
-        proportions = counts.float() / len(y)
+        # create self.n_pseudobulks
+        if self.n_cells_per_pseudobulk is None or self.n_cells_per_pseudobulk > x_.shape[0] :
+            n_cells_per_pseudobulk = x_.shape[0]
+        else:
+            n_cells_per_pseudobulk = self.n_cells_per_pseudobulk
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py:324]: Why duplicate this variable ?

File: scvi/module/_mixupvae.py [hunk 6]
```diff
@@ -287,42 +313,57 @@ def inference(
...
+        random_state = np.random.RandomState(seed=42)
+        pseudobulk_indices = random_state.choice(
+            x_.shape[0],
+            size = (self.n_pseudobulks, n_cells_per_pseudobulk),
+            replace = True,
+        )
+        x_pseudobulk_ = x_[pseudobulk_indices, :].mean(axis=1)
+
+        y_ = y[pseudobulk_indices, :].squeeze(axis=2)
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py]: suggestion: rename to `y_pseudobulk` to avoid confusion. (`x_` refers to the single-cell batch)

File: benchmark_utils/tuning_utils.py (hunk 3 above)
Comment (SimonGrouard) [reply benchmark_utils/tuning_utils.py]: Yes I've changed it to "-1", thanks! Weird that you got an issue with it, when running it on my side, the 6th part is always the last one, so no problem...
Agreed that this way is not ideal overall, but I'm not sure how to get these information otherwise!

File: scvi/module/_mixupvae.py [hunk 7]
```diff
@@ -336,30 +377,28 @@ def inference(
...
                     categorical_input.append(one_hot_cat_covs)
-                    ont_hot_cat_covs_pure = one_hot_cat_covs.mean(axis=0).resize(1,n_cat)
-                    categorical_pseudobulk_input.append(ont_hot_cat_covs_pure)
-                    categorical_signature_input_temp = []
-                    for cell_type in unique_indices:
-                        idx = (y == cell_type).flatten()
-                        one_hot_cat_covs_pure = one_hot_cat_covs[idx, :].mean(axis=0)
-                        categorical_signature_input_temp.append(one_hot_cat_covs_pure)
-                    categorical_signature_input.append(torch.stack(categorical_signature_input_temp))
+                    one_hot_cat_covs_pure = one_hot_cat_covs[pseudobulk_indices, :].mean(axis=1)
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py]: should be updated by merging with main.

File: scvi/module/_mixupvae.py [hunk 8]
```diff
@@ -287,42 +313,57 @@ def inference(
...
+        x_signature_mask = []
+        unique_indices, counts = y.unique(return_counts=True)
         for cell_type in unique_indices:
             idx = (y == cell_type).flatten()
-            x_pure = x_[idx, :].mean(axis=0)
-            x_signature_.append(x_pure)
-        x_signature_ = torch.stack(x_signature_, dim=0)
+            x_signature_mask.append(idx.tolist())
+        x_signature_mask = torch.Tensor(x_signature_mask).to(device="cuda")
+        x_signature_ = torch.matmul(x_signature_mask, x_)/counts.unsqueeze(-1)
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py:338]: we could also consider moving this to utils for readability. Maybe in another PR if this is too much.

Comment (khalilouardini) [reply scvi/module/_mixupvae.py:338]: sanity check: did you check that the signature results were the same ?

Comment (khalilouardini) [reply scvi/module/_mixupvae.py:338]: nvm I did I get the same result

File: scvi/module/_mixupvae.py [hunk 9]
```diff
@@ -435,18 +469,21 @@ def generative(
...
+        pseudobulk_indices=(),
         size_factor=None,
         y=None,
         transform_batch=None,
     ):
         """Runs the generative model."""
         if self.pseudo_bulk == "post_inference":
             # create it from z by overwritting the one coming from inference
-            z_pseudobulk = z.mean(axis=0).unsqueeze(0)
+            z_pseudobulk = z[pseudobulk_indices, :].mean(axis=1)
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py:491]: I don"t remember the reason why we are overwriting `z_pseudobulk` here ..

File: scvi/module/_mixupvae.py [hunk 10]
```diff
@@ -479,14 +515,16 @@ def generative(
...
             size_factor,
-            batch_index,
+            one_hot_batch_index,
             *categorical_input,
-            y,
+            one_hot_y,
         )
+        one_hot_y_pseudobulk = one_hot_y[pseudobulk_indices, :].mean(axis=1)
         (
             px_pseudobulk_scale,
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py:534]: suggestion: we can add an argument like `reconstruct_pseudobulk` set to `False` by default. There's no need to run the decoder on the pseudobulk if we don't add to the loss. Or we can use the `loss_computation` attribute of the class directly

File: scvi/module/_mixupvae.py [hunk 11]
```diff
@@ -496,17 +534,20 @@ def generative(
             self.dispersion,
             decoder_pseudobulk_input,
             size_factor_pseudobulk,
-            batch_index,
+            one_hot_batch_index_pseudobulk,
             *categorical_pseudobulk_input,
-            y,
-        )  # for pseudobulk, batch_index, categorical_input and y are not the right dimension
+            one_hot_y_pseudobulk,
```
Comment (khalilouardini) [inline scvi/module/_mixupvae.py:544]: Just to confirm again: the labels are not decoded right ?

**Review by khalilouardini — COMMENTED** [review]
These comments are about the changes related to the model! Thanks Simon

One sanity check, have you checked that training was still consistent with what we've seen before with all these changes ?

File: scvi/module/_mixupvae.py (hunk 2 above)
Comment (khalilouardini) [reply scvi/module/_mixupvae.py:182]: I think we talked about this but the reconstruted loss by be huge because the decoder is not trying to fit the pseudobulk data (no backprop) - so we can try learning w/ the  pseudobulk reconstruction and see if the problem persists. I'll look into that after we merge this.

Comment (SimonGrouard) [reply scvi/module/_mixupvae.py:182]: Well it's also because by summing across the variable dimension gives a huge number if the variable dimension is big (eg in the input space). But yes to investigate

File: scvi/module/_mixupvae.py (hunk 9 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py:491]: It's because if we want to create the pseudobulk post-inference (and not pre-encoded), then we cannot use the one coming from the inference outputs, we want to re-generate it from the z coming from the inference outputs. Does it make sense now ?

File: scvi/module/_mixupvae.py (hunk 11 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py:544]: You're right, they are not decoded

File: scvi/module/_mixupvae.py (hunk 5 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py:324]: If the user set `self.n_cells_per_pseudobulk` to be None, or to be greater than the number of cells in a batch, then it will be set to the size of the batch. 
If `self.n_cells_per_pseudobulk` is smaller than number of cells in the batch, then good, keep it as it is.

File: scvi/module/_mixupvae.py (hunk 10 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py:534]: The issue is that there are a lot of if statements using the pseudobulk, so needing another if statements makes the code way harder to read, because we have to add them a lot of time

File: scvi/module/_mixupvae.py (hunk 8 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py:338]: I moved it to utils in the new commit

File: scvi/module/_mixupvae.py (hunk 7 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py]: Done

File: scvi/module/_mixupvae.py (hunk 6 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py]: Done

File: scvi/module/_mixupvae.py (hunk 4 above)
Comment (SimonGrouard) [reply scvi/module/_mixupvae.py]: Moved to utils

**Review by khalilouardini — APPROVED** [review]
